ocrd_eval.md

```diff
@@ -0,0 +1,372 @@
+
+At this point we only provide a CER per page; an overall CER might be calculated as a weighted aggregate at a later stage.
+
+##### Word Error Rate (WER)
+
+The word error rate (WER) is closely connected to the CER. While the CER focusses on differences between characters, the WER represents the percentage of words incorrectly recognized in a text.
+
+CER and WER share categories of errors, and the WER is similarly calculated:
+
+$WER = \frac{i_w + s_w + d_w}{n_w}$
+
+where $i_w$ is the number of inserted, $s_w$ the number of substituted, $d_w$ the number of deleted and $n_w$ the total number of words.
+
+More specific cases of WER consider only the "significant" words, omitting e.g. stopwords from the calculation.
+
+###### WER Granularity
+
+In OCR-D we distinguish between the WER per **page** and the **overall** WER of a text. The reasoning here follows the one of CER granularity.
+
+At this point we only provide a WER per page; an overall WER might be calculated at a later stage.
+
+#### Bag of Words
+
+In the "Bag of Words" model a text is represented as a set of its word irregardless of word order or grammar; Only the words themselves and their number of occurence are considered.
+
+Example:
+
+![A sample paragraph in German Fraktur](https://pad.gwdg.de/uploads/4d33b422-6c77-436c-a3e6-bf27e67dc203.jpg)
+
+> Eine Mondfinsternis ist die Himmelsbegebenheit welche sich zur Zeit des Vollmondes ereignet, wenn die Erde zwischen der Sonne und dem Monde steht, so daß die Strahlen der Sonne von der Erde aufgehalten werden, und daß man so den Schatten der Erde in dem Monde siehet. In diesem Jahre sind zwey Monfinsternisse, davon ist ebenfalls nur Eine bey uns sichtbar, und zwar am 30sten März des Morgens nach 4 Uhr, und währt bis nach 6 Uhr.
+
+To get the Bag of Words of this paragraph a set containing each word and its number of occurence is created:
+
+$BoW$ =
+
+```json=
+{
+    "Eine": 2, "Mondfinsternis": 1, "ist": 2, "die": 2, "Himmelsbegebenheit": 1, 
+    "welche": 1, "sich": 1, "zur": 1,  "Zeit": 1, "des": 2, "Vollmondes": 1,
+    "ereignet,": 1, "wenn":1, "Erde": 3, "zwischen": 1, "der": 4, "Sonne": 2,
+    "und": 4, "dem": 2, "Monde": 2, "steht,": 1, "so": 2, "daß": 2, 
+    "Strahlen": 1, "von": 1, "aufgehalten": 1, "werden,": 1, "man": 1, "den": 1, 
+    "Schatten": 1, "in": 1, "siehet.": 1, "In": 1, "diesem": 1, "Jahre": 1, 
+    "sind": 1, "zwey": 1, "Monfinsternisse,": 1, "davon": 1, "ebenfalls": 1, "nur": 1, 
+    "bey": 1, "uns": 1, "sichtbar,": 1, "zwar": 1, "am": 1, "30sten": 1, 
+    "März": 1, "Morgens": 1, "nach": 2, "4": 1, "Uhr,": 1, "währt": 1, 
+    "bis": 1, "6": 1, "Uhr.": 1
+}
+```
+
+### Layout Evaluation
```

Comment:
Thank you for your thoughts.

**ad 1:** That's something I will have to talk about with @cneud and @kba .

**ad 2:** This is true. We can rephrase the introduction in way that makes this clear but still emphasizes that this especially problematic for documents with complex layout.

**ad 3:** 

> if coordinates are inaccurate locally (too wide around the glyphs for the OCR engine to still catch them adequately, or too narrow, cutting off foreground)

Where do you draw the line between layout and character level? Or is there no such distinction and single characters are also layout?

**ad 4:** I haven't read anything about a metric for reading order. Seems like I have blind spot. Could you point me to some paper where this is done?

**ad 5:** 
* that's right, we should be more precise about the level
* aggregating over segments is implicitly meant since we deal with layout (see question at 3)
* regarding the other points you make: these are things I haven't thought about so far, true, so thanks for pointing them out. I will have to research a bit and come back to these points.

Reply:
> **ad 3:**
> 
> > if coordinates are inaccurate locally (too wide around the glyphs for the OCR engine to still catch them adequately, or too narrow, cutting off foreground)
> 
> Where do you draw the line between layout and character level? Or is there no such distinction and single characters are also layout?

Neither of those. I did not touch character level. My concern is with line level (i.e. textline segmentation quality). The existing formulations sometimes seem too specific to region level segmentation (is all I'm saying).


> **ad 4:** I haven't read anything about a metric for reading order. Seems like I have blind spot. Could you point me to some paper where this is done?

Might well be a blind spot, I don't recall anything. (But it should not be difficult to device a suitable metric – I'll get back to you with a better answer.)

> **ad 5:**
> 
>     * that's right, we should be more precise about the level

better even: provide both (region and line IoU perspective). (In ocrd-segment-evaluate, I provide both region and line level; PRImA does that, too.)

>     * aggregating over segments is implicitly meant since we deal with layout (see question at 3)

ok, but then you miss a neutral metric that is independent over how many segments there are on a page and how large they are. (In ocrd-segment-evaluate, I provide both pixel-wise and segment-wise aggregates; PRImA does that, too.)

>     * regarding the other points you make: these are things I haven't thought about so far, true, so thanks for pointing them out. I will have to research a bit and come back to these points.

Let's have a dedicated VC session on that, we could look at ocrd-segment-evaluate for example.

Reply:
> the reasoning seems inverse: We don't have "text accuracy obstructed by segmentation quality", but "segmentation as a prerequisite to even be able to measure text accuracy". So either you say
> 
>     * "both problems must be disentangled as much as possible"
>       → and go for OCR-GT text evaluation on GT segments on the one side, bundled with OCR-GT layout evaluation.
>       Or you say
> 
>     * "I want to get a realistic overall text measure, but disregarding reading order" (i.e. total order of lines, so misdetected columns and other subline-level disorders would still be seen as errors)
>       → and go for measures like Flexible Character Accuracy

Result of our discussion yesterday is that we opt for your second approach. I'll rephrase the paragraph to make it a bit clearer.

Reply:
> > **ad 3:**
> > > if coordinates are inaccurate locally (too wide around the glyphs for the OCR engine to still catch them adequately, or too narrow, cutting off foreground)
> > 
> > 
> > Where do you draw the line between layout and character level? Or is there no such distinction and single characters are also layout?
> 
> Neither of those. I did not touch character level. My concern is with line level (i.e. textline segmentation quality). The existing formulations sometimes seem too specific to region level segmentation (is all I'm saying).

@cneud, @kba and I talked about this yesterday. Admittedly, line level has it's place, but from our pragmatic point of view region level is sufficient.

Reply:
> > **ad 4:** I haven't read anything about a metric for reading order. Seems like I have blind spot. Could you point me to some paper where this is done?
> 
> Might well be a blind spot, I don't recall anything. (But it should not be difficult to device a suitable metric – I'll get back to you with a better answer.)

We came up with this papier: https://www.primaresearch.org/www/assets/papers/ICDAR2013_Clausner_ReadingOrder.pdf.

The "Reading Order" section was more about providing a bit of background than actually defining a metric (which would be a part of our MVP), but we can add this in the "not yet to be implemented" section.

Reply:
Thanks for pointing me to that! Very interesting. 

We should definitely mention this – if for the sake of completeness / commensurability.

I am surprised to find the actual formula slightly arbitrary, though. In trying to come up with a **rate**, the authors say:
> it is favourable to calculate a relative error or success measure in the form of a percentage. This can be achieved by relating the error value to the highest possible error value. Due to the unconstrained nature of layout analysis results a definitive maximum cannot be determined. There is for instance no limit to the number of overlapping/stacked regions. Instead, a non-linear success function is used which has a parameter ($e_50$) representing an error value that corresponds to a success rate of 50%.
   
IMO it would be natural to use the share/number of **pixels** of each overlap area as weight. Then no such non-linear term would be necessary (the denominator would be the overall size of the page, times the sum of possible penalties)...

Reply:
See f8175210c93498ffa476d98816471cf63428fb68 for reading order evaluation.

Reply:
> IMO it would be natural to use the share/number of pixels of each overlap area as weight. Then no such non-linear term would be necessary (the denominator would be the overall size of the page, times the sum of possible penalties)...

Do I sense a new paper coming up…? ;)

Reply:
Resolved?

Reply:
> Resolved?

not quite – see latest review

Reply:
> > IMO it would be natural to use the share/number of pixels of each overlap area as weight. Then no such non-linear term would be necessary (the denominator would be the overall size of the page, times the sum of possible penalties)...
> 
> Do I sense a new paper coming up…? ;)

Another time, perhaps. When we get to re-implement PRImA Layout Eval in Open Source, we'll have to revisit (and communicate with Christian) anyway.

Reply:
I'll transfer your comment to a ticket then so we don't forget.

Reply:
See https://github.com/OCR-D/spec/issues/238. AFAICT I accepted all your changes dealing with reading order. Please `resolve` if you see fit. :)